sim/team-validator.ts

```diff
@@ -9,6 +9,155 @@
 
 import {Dex} from './dex';
 
+/**
+ * Describes a possible way to get a pokemon. Is not exhaustive!
+ * sourcesBefore covers all sources that do not have exclusive
+ * moves (like catching wild pokemon).
+ *
+ * First character is a generation number, 1-7.
+ * Second character is a source ID, one of:
+ *
+ * - E = egg, 3rd char+ is the father in gen 2-5, empty in gen 6-7
+ *   because egg moves aren't restricted to fathers anymore
+ * - S = event, 3rd char+ is the index in .eventPokemon
+ * - D = Dream World, only 5D is valid
+ * - V = Virtual Console transfer, only 7V is valid
+ *
+ * Designed to match MoveSource where possible.
+ */
+export type PokemonSource = string;
+
+/**
+ * Represents a set of possible ways to get a Pokémon with a given
+ * set.
+ *
+ * `new PokemonSources()` creates an empty set;
+ * `new PokemonSources(dex.gen)` allows all Pokemon.
+ */
+export class PokemonSources {
+	/**
+	 * A list of specific possible PokemonSources
+	 */
+	sources: PokemonSource[];
```

Comment:
It's not a Set mostly for perf reasons: the size of this is usually 0-5, and arrays are faster.

Checking last-element-only is actually enough to guarantee uniqueness (in the old code, there's only one place where you even need to check it), but also uniqueness isn't actually important for many use-cases.

Reply:
Well, I'm not sure I'd call it perf reasons; the real reason is because `Set`s didn't exist at the time this code was written in 2011, and I never updated it. Maybe it might be worth changing it to Set and benchmarking. Can you?

Reply:
I already checked small arrays a while back - youre right, for sizes less than even like 8-10 its still more performant to use arrays. However, because this is a little confusing to call it a set and because its an optimization, can you please add comments to these effects?